Queues - Kayla Kubicke - js-scrabble

moonpastry opened the pull request:
# JS Scrabble
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | Any time I learn something, it changes the way I approach things in the future so it is hard to answer this question. I think the googling skills I learned during ruby were particularly helpful during this project.   |
| What was a challenge you were able to overcome on this assignment? | I struggled with the prototype function highestScoreFrom. I figured it out but if I had time I would refactor the function. |
| What is your favorite thing about learning a new programming language? | I always enjoy problem solving when learning a new language. |
| What is your least favorite thing about learning a new programming language? | Getting stuck on a small error because you don't have enough knowledge to quickly debug.  |
| Do you have any recommendations on how we could improve this project for the next cohort? |  |

Commit:
Scrabble word returns a score.

A word score's is calculated based on individual letters.

Commit:
Merge pull request #1 from moonpastry/score

Scrabble word returns a score.

Commit:
Scrabble words are sorted based on highest score.

An array can be passed through a function and words are sorted based
on their individual score.

Commit:
If last word is the highest scoring word, word is returned.

In the simple case that the last word is the highest scoring, that word
is returned.
This function runs before the others to avoid running through
unnecessary operations.

Commit:
An array of highest earning words is created.

If multiple words have the same high earning score, an array collection is created.
This array can be used to evaluate the length of the highest earning words.

Commit:
Highest earning score rules are applied.

A winning word can be chosen among other high earning words.
The first 7 letter word is automatically the winner.
If none of the words are 7 letters, the shortest word with the
lowest index is chosen.

Commit:
Merge pull request #2 from moonpastry/highestScore

Highest score

Commit:
Created player constructor object.

Player object has a name, an array that will contain any words they
play, and creates a new Scrabble object so that prototype functions
can be called on either a word or an array of words.

Commit:
Player can add word to word collection array.

Added prototype function that adds an individual word to
word collection array.
Removed notes and cleaned up scrabble.js.

Commit:
Wrote skeleton for totalScore.

Commit:
A player's total word score is returned.

Prototype function totalScore grabs an array of words, calculates
the total score, and returns the total score.

Commit:
Player can determine if they won the game based on score.

Prototype function hasWon returns true or false based on
player's score.
Prototype function play returns false if a player has already won the game.

Commit:
Player's highest scoring word is returned.

Commit:
Highest word score is returned.

Commit:
Merge pull request #3 from moonpastry/playerobject

Playerobject

Commit:
DRYed up score function.

Worked with Noam to reduce amount of code in score function.

Commit:
Transformed string to array so score can be calculated.

Commit:
Refactored highestScoreFrom.

Reduced amount of code used to calculate the highest scoring word.

Commit:
Merge pull request #4 from moonpastry/hsfrefactor

Hsfrefactor

Commit:
Scrabble and Player object interact properly.

Commit:
Removed testing code.

Commit:
Merge pull request #5 from moonpastry/playerFix

Player fix

Commit:
Refactored code based on Noam's suggestions.

Removed sort loop from highestScoreFrom as it was unnecessary.
Removed initial value from reduce loop in highestScoreFrom because it will grab the first element as it's initial value.
Added rules and changed variable names to increase readability.
Cleaned up logic.
Fixed highestScoreFrom bug; first 7 letter word will be returned.

Commit:
Merge pull request #6 from moonpastry/refactor

Refactor

PilgrimMemoirs commented:
# JS Scrabble
## What We're Looking For

| Feature | Feedback |
| :------------- | :------------- |
| **Core Requirements** | |
| Git hygiene | Well Done |
| Comprehension questions	| Well Done |
| **General** | |
| `score` calculates score, has appropriate params and return value | Well Done |
| `highestScoreFrom` calculates highest scoring word, has appropriate params and return value | Well done |
| **`Player` object** | |
| Has `name` and `plays` properties | Well Done |
| Has `play`, `totalScore`, `hasWon` functions | Well Done |
| Has `highestScoringWord` and `highestWordScore` functions | Well Done |
| **Overall** | |
| | Submission is complete and meets expectations of this project. Great work! |

Commit:
Fixed two bugs.

Reintroduced 'i' into regular expressions based on Noam's suggestions. (I misunderstood their purpose.)
First shortest word with the highest score is returned in cases where the winning word is not seven characters long and multiple words have the same length and score.

Commit:
Merge pull request #7 from moonpastry/refactor2

Fixed two bugs.
